python/lsst/daf/butler/core/dbAuth.py

```diff
@@ -0,0 +1,216 @@
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import fnmatch
+import os
+import stat
+import urllib.parse
+import yaml
+
+__all__ = ["DbAuth", "DbAuthError"]
+
+
+class DbAuthError(RuntimeError):
+    """A problem has occurred retrieving database authentication information.
+    """
+    pass
+
+
+class DbAuth:
+    """Retrieves authentication information for database connections.
+
+    The authorization configuration is taken from the ``authList`` parameter
+    or a (group- and world-inaccessible) YAML file located at a path specified
+    by the given environment variable or at a default path location.
+
+    Parameters
+    ----------
+    path : `str` or None, optional
+        Path to configuration file.
+    envVar : `str` or None, optional
+        Name of environment variable pointing to configuration file.
+    authList : `list` [`dict`] or None, optional
+        Authentication configuration.
+
+    Notes
+    -----
+    At least one of ``path``, ``envVar``, or ``authList`` must be provided;
+    generally ``path`` should be provided as a default location.
+    """
+    def __init__(self, path=None, envVar=None, authList=None):
+        if authList is not None:
+            self.authList = authList
+            return
+        if envVar is not None and envVar in os.environ:
+            secretPath = os.path.expanduser(os.environ[envVar])
+        elif path is None:
+            raise DbAuthError(
+                "No default path provided to DbAuth configuration file")
+        else:
+            secretPath = os.path.expanduser(path)
+        if not os.path.isfile(secretPath):
+            raise DbAuthError("No DbAuth configuration file: " + secretPath)
+        mode = os.stat(secretPath).st_mode
+        if mode & (stat.S_IRWXG | stat.S_IRWXO) != 0:
+            raise DbAuthError(
+                "DbAuth configuration file {} has incorrect permissions: "
+                "{:o}".format(secretPath, mode))
+
+        try:
+            with open(secretPath) as secretFile:
+                self.authList = yaml.safe_load(secretFile)
+        except Exception as exc:
+            raise DbAuthError(
+                "Unable to load DbAuth configuration file: " +
+                secretPath) from exc
+
+    def getAuth(self, drivername, username, host, port, database):
```

Comment:
Could some parameters made default, I'd prefer not to provide port number if I do not know it?

Reply:
Another idea - could this method have an overload which takes url string and returns new string with username and password in it, I think it would be even more usable.

Reply:
The idea was that the arguments are the components of the SQLAlchemy URL object.  Making parameters defaultable would I think have to make the order different than in the URL and less memorable.

Reply:
(And you don't have to provide the port number; None or "" are fine.)

If I were going to take a string URL, I'd make that a separate method.  That would be easy enough to do if it's useful.

Returning a re-composed URL from a new URL-taking method shouldn't be too hard.  It's not possible with the current method, as all parts of the URL aren't available.

Is anyone going to use such a new method to make it worthwhile writing the tests?

Reply:
If this is purely sqlalchemy-oriented then it may be more natural to accept `URL` instance and return modified `URL`. I was thinking about other possible use cases (e.g. PPDB with Cassandra) which do not use sqlalchemy at all but could still mimic URL-style connection string, but given that this code is in `daf_butler` it is probably not very reusable for other clients anyways.

Reply:
We could in theory move it into a dbutils package if needs be.

Reply:
While it is SQLAlchemy-oriented, I didn't want it to actually depend on SQLAlchemy.  I was hoping that others could reuse it in the future, but it wasn't clear that we have other immediate clients.